# Get HTML Table

In [1]:
pip install lxml

Note: you may need to restart the kernel to use updated packages.


In [2]:
import requests
import lxml.html as lh
import pandas as pd

In [3]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
#Create a handle, page, to handle the contents of the website
page = requests.get(url)
#Store the contents of the website under doc
doc = lh.fromstring(page.content)
#Parse data that are stored between <tr>..</tr> of HTML
tr_elements = doc.xpath('//tr')

In [4]:
#Check the length of the first 12 rows
[len(T) for T in tr_elements[:12]]

[3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

### Next, let’s parse the first row as our header.

In [5]:
tr_elements = doc.xpath('//tr')
#Create empty list
col=[]
i=0
#For each row, store each first element (header) and an empty list
for t in tr_elements[0]:
    i+=1
    name=t.text_content()
    print('%d %s' % (i, name))
    col.append((name,[]))

1 Postal Code

2 Borough

3 Neighborhood



## Creating Pandas DataFrame

In [6]:
#Since out first row is the header, data is stored on the second row onwards
for j in range(1,len(tr_elements)):
    #T is our j'th row
    T=tr_elements[j]
    
    #If row is not of size 10, the //tr data is not from our table 
    if len(T)!=3:
        break
    
    #i is the index of our column
    i=0
    
    #Iterate through each element of the row
    for t in T.iterchildren():
        data=t.text_content() 
        #Check if row is empty
        if i>0:
        #Convert any numerical value to integers
            try:
                data=int(data)
            except:
                pass
        #Append the data to the empty list of the i'th column
        col[i][1].append(data)
        #Increment i for the next column
        i+=1

In [7]:
print(T)

<Element tr at 0x7f3fd3fd8ea8>


In [8]:
# Just to be sure, let’s check the length of each column. Ideally, they should all be the same.
[len(C) for (title,C) in col]

[181, 181, 181]

#### Perfect! This shows that each of our 3 columns has exactly 181 values.

In [9]:
#Now we are ready to create the dataframe:
Dict={title:column for (title,column) in col}
df=pd.DataFrame(Dict)
df.head()

Postal Code\n           Borough\n               Neighborhood\n
0         M1A\n      Not assigned\n               Not assigned\n
1         M2A\n      Not assigned\n               Not assigned\n
2         M3A\n        North York\n                  Parkwoods\n
3         M4A\n        North York\n           Victoria Village\n
4         M5A\n  Downtown Toronto\n  Regent Park, Harbourfront\n

In [10]:
df.columns = ['PostalCode', 'Borough', 'Neighborhood']
df = df.replace(to_replace ='\n', value = '', regex = True) 
df.head()

PostalCode           Borough               Neighborhood
0        M1A      Not assigned               Not assigned
1        M2A      Not assigned               Not assigned
2        M3A        North York                  Parkwoods
3        M4A        North York           Victoria Village
4        M5A  Downtown Toronto  Regent Park, Harbourfront

#### Ignore the Not Assigned Borough

In [11]:
df = df[df.Borough != 'Not assigned']
df.head()

PostalCode           Borough                                 Neighborhood
2        M3A        North York                                    Parkwoods
3        M4A        North York                             Victoria Village
4        M5A  Downtown Toronto                    Regent Park, Harbourfront
5        M6A        North York             Lawrence Manor, Lawrence Heights
6        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [12]:
df.loc[df['Neighborhood'] == "Not assigned", 'Neighborhood'] = df['Borough']
df.reset_index()

index PostalCode                Borough  \
0        2        M3A             North York   
1        3        M4A             North York   
2        4        M5A       Downtown Toronto   
3        5        M6A             North York   
4        6        M7A       Downtown Toronto   
..     ...        ...                    ...   
99     165        M4Y       Downtown Toronto   
100    168        M7Y           East Toronto   
101    169        M8Y              Etobicoke   
102    178        M8Z              Etobicoke   
103    180             Canadian postal codes   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  
103                                                     

[104 rows x 4 columns]

In [13]:
df.head()

PostalCode           Borough                                 Neighborhood
2        M3A        North York                                    Parkwoods
3        M4A        North York                             Victoria Village
4        M5A  Downtown Toronto                    Regent Park, Harbourfront
5        M6A        North York             Lawrence Manor, Lawrence Heights
6        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [14]:
df.shape

(104, 3)

In [15]:
geolocation = pd.read_csv("http://cocl.us/Geospatial_data")
geolocation.columns = ['PostalCode', 'Latitude', 'Longitude']
geolocation.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

In [16]:
df_WithGeoLoc = pd.merge(df, geolocation, how='left', on=['PostalCode'])
df_WithGeoLoc.dropna()

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
..         ...               ...   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [17]:
df_WithGeoLoc.groupby('PostalCode')
df_WithGeoLoc

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A       Downtown Toronto   
..         ...                    ...   
99         M4Y       Downtown Toronto   
100        M7Y           East Toronto   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   
103             Canadian postal codes   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  
103                                                           NaN        NaN  

[104 rows x 5 columns]

## Creating a copy of df and filtering only North York Borough

In [18]:
dfNY = df_WithGeoLoc.loc[(df_WithGeoLoc.Borough == "North York")]
dfNY.head()

PostalCode     Borough                      Neighborhood   Latitude  \
0         M3A  North York                         Parkwoods  43.753259   
1         M4A  North York                  Victoria Village  43.725882   
3         M6A  North York  Lawrence Manor, Lawrence Heights  43.718518   
7         M3B  North York                         Don Mills  43.745906   
10        M6B  North York                         Glencairn  43.709577   

    Longitude  
0  -79.329656  
1  -79.315572  
3  -79.464763  
7  -79.352188  
10 -79.445073

In [22]:
#NorthYork center
centerLat = 43.761539
centerLong = -79.411079

## Creating the Map of North York with the Center in red and all the other point in blue

In [23]:
import folium

In [38]:
venues_map = folium.Map(location=[centerLat, centerLong], zoom_start=10) # generate map centred around the Conrad Hotel

# add a red circle marker to represent the Conrad Hotel
folium.features.CircleMarker(
    [centerLat, centerLong],
    radius=10,
    color='red',
    popup='Center',
    fill = True,
    fill_color = 'red',
    fill_opacity = 0.6
).add_to(venues_map)

# add the othre points as blue circle markers
for lat, lng, label in zip(dfNY.Latitude, dfNY.Longitude, dfNY.Borough):
    folium.features.CircleMarker(
        [lat, lng],
        radius=5,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(venues_map)

# display map
venues_map

## Creating the Map of Toronto with the Center in red and all the other point in blue

In [28]:
dfToronto = df_WithGeoLoc.loc[(df_WithGeoLoc.Borough != "North York")]
dfToronto = dfToronto.dropna()

In [29]:
#Toronto center
centerLatT = 43.653908
centerLongT = -79.384293

In [35]:
dfToronto.head()

PostalCode           Borough                                 Neighborhood  \
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
5        M9A         Etobicoke      Islington Avenue, Humber Valley Village   
6        M1B       Scarborough                               Malvern, Rouge   
8        M4B         East York              Parkview Hill, Woodbine Gardens   

    Latitude  Longitude  
2  43.654260 -79.360636  
4  43.662301 -79.389494  
5  43.667856 -79.532242  
6  43.806686 -79.194353  
8  43.706397 -79.309937

In [37]:
toronto_map = folium.Map(location=[centerLatT, centerLongT], zoom_start=10) # generate map centred around the Conrad Hotel

# add a red circle marker to represent the Conrad Hotel
folium.features.CircleMarker(
    [centerLatT, centerLongT],
    radius=10,
    color='red',
    popup='Center',
    fill = True,
    fill_color = 'red',
    fill_opacity = 0.6
).add_to(toronto_map)

# add the othre points as blue circle markers
for lat, lng, label in zip(dfToronto.Latitude, dfToronto.Longitude, dfToronto.Borough):
    folium.features.CircleMarker(
        [lat, lng],
        radius=5,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(toronto_map)


# display map
toronto_map

## Merging North York and Toronto and showing on Map

In [40]:
dfAll = dfToronto.append(dfNY)
dfAll.shape

(103, 5)

In [42]:
All_map = folium.Map(location=[centerLatT, centerLongT], zoom_start=10) # generate map centred around the Conrad Hotel

# add the othre points as blue circle markers
for lat, lng, label in zip(dfAll.Latitude, dfAll.Longitude, dfAll.Borough):
    folium.features.CircleMarker(
        [lat, lng],
        radius=5,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(All_map)


# display map
All_map

# Using K-Means Algorithm

In [43]:
from sklearn.cluster import KMeans 

In [44]:
k_means = KMeans(init="k-means++", n_clusters=4, n_init=12)

In [45]:
X = dfAll[['Latitude', 'Longitude']]

In [47]:
k_means.fit(X)

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
    n_clusters=4, n_init=12, n_jobs=None, precompute_distances='auto',
    random_state=None, tol=0.0001, verbose=0)

In [48]:
k_means_labels = k_means.labels_
k_means_labels

array([2, 2, 3, 1, 2, 2, 3, 1, 2, 2, 0, 3, 1, 2, 2, 0, 1, 2, 2, 2, 1, 2,
       2, 2, 1, 2, 2, 2, 1, 2, 2, 2, 1, 2, 2, 1, 2, 3, 1, 0, 0, 3, 3, 1,
       0, 2, 3, 3, 1, 0, 2, 3, 3, 3, 1, 2, 2, 3, 1, 2, 2, 1, 2, 2, 3, 3,
       1, 2, 2, 3, 3, 1, 2, 2, 3, 2, 2, 3, 3, 1, 1, 0, 0, 0, 2, 0, 0, 0,
       0, 0, 0, 0, 3, 3, 3, 0, 3, 0, 3, 0, 3, 0, 0], dtype=int32)

#### Adding K_mean labels in the dataframe

In [62]:
dfAll['Kvalue'] = k_means_labels
dfAll.head()

PostalCode           Borough                                 Neighborhood  \
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
5        M9A         Etobicoke      Islington Avenue, Humber Valley Village   
6        M1B       Scarborough                               Malvern, Rouge   
8        M4B         East York              Parkview Hill, Woodbine Gardens   

    Latitude  Longitude  Kvalue  
2  43.654260 -79.360636       2  
4  43.662301 -79.389494       2  
5  43.667856 -79.532242       3  
6  43.806686 -79.194353       1  
8  43.706397 -79.309937       2

In [53]:
k_means_cluster_centers = k_means.cluster_centers_

## Map creation:
- Centroids in red
- 4 clusters divided by colors

In [72]:
All_Kmap = folium.Map(location=[centerLatT, centerLongT], zoom_start=10) # generate map centred around the Conrad Hotel

# add a red circle marker to represent the Conrad Hotel
for i in k_means_cluster_centers:
    folium.features.CircleMarker(
        i,
        radius=5, 
        color='red',
        fill = True,
        fill_color = 'red',
        # fill_opacity = 0.6
    ).add_to(All_Kmap)
    
# add the othre points as blue circle markers
for lat, lng, label, k in zip(dfAll.Latitude, dfAll.Longitude, dfAll.Borough, dfAll.Kvalue):
    if k == 0:
        newColor='yellow'
    if k == 1:
        newColor='green'
    if k == 2:
        newColor='gray'
    if k == 3:
        newColor='orange'
    folium.features.CircleMarker(
        [lat, lng],
        radius=5,
        color=newColor,
        popup=label,
        fill = True,
        fill_color=newColor,
        fill_opacity=0.6
    ).add_to(All_Kmap)

All_Kmap